### ISSUE TITLE
TimezoneRepositoryProviderSPI NullPointerException

### ISSUE THREAD

**Issue by cp949** [MenoData/Time4J#464]
My environment
java 8 and tomcat 8
no custom properties
    (ex: 'net.time4j.tz.repository.path' or 'net.time4j.tz.repository.version' )

build.gradle
	// time4j
	compile ("net.time4j:time4j-olson:4.12")
	compile ("net.time4j:time4j-tzdata:1.7-2016a")
	compile ("net.time4j:time4j-i18n:4.12")
	compile ("net.time4j:time4j-calendar:4.12")
	compile ("net.time4j:time4j-range:4.12")
	compile ("net.time4j:time4j-misc:4.12")
	compile ("net.time4j:time4j-core:4.12")

exception logs
NullPointerException when start
java.lang.NullPointerException
at java.io.DataInputStream.readByte(DataInputStream.java:265)
at net.time4j.tz.spi.TimezoneRepositoryProviderSPI.checkMagicLabel(TimezoneRepositoryProviderSPI.java)
at net.time4j.tz.spi.TimezoneRepositoryProviderSPI.<init>(TimezoneRepositoryProviderSPI.java)

**Comment by MenoData** [comment]
Hm, I cannot reproduce your problem using my own Tomcat-setup. Anyway, it is a bug. Unfortunately there is no active error logging so I am definitely going to improve logging in next release v4.13. In the meantime:

**May I kindly ask you for trying out following special setup to enable a better diagnosis just now:**

a) compile following class and put in into your class path (the same where your time4j-jars are)

```java
package net.time4j.base;

import java.io.IOException;
import java.io.InputStream;
import java.net.URI;
import java.net.URISyntaxException;
import java.net.URL;
import java.net.URLConnection;
import java.security.CodeSource;
import java.security.ProtectionDomain;
import java.util.ServiceLoader;

public class DebugResourceLoader
    extends ResourceLoader {

    @Override
    public URI locate(
        String moduleName,
        Class<?> moduleRef,
        String path
    ) {
        // first try uri construction whose initialization time is much quicker than querying the class loader
        String constructedUri = null;

        try {
            ProtectionDomain pd = moduleRef.getProtectionDomain(); // null on android platform
            CodeSource cs = ((pd == null) ? null : pd.getCodeSource());

            if (cs != null) {
                constructedUri = cs.getLocation().toExternalForm();
                if (constructedUri.endsWith(".jar")) {
                    constructedUri = "jar:" + constructedUri + "!/";
                }
                constructedUri += path;
                return new URI(constructedUri);
            }
        } catch (SecurityException se) {
            // use fallback via class loader
        } catch (URISyntaxException e) {
            System.err.println("Warning: malformed resource path = " + constructedUri);
        }

        try {
            // last try - ask the class loader
            URL url = moduleRef.getClassLoader().getResource(path);
            if (url != null) {
                return url.toURI();
            }
        } catch (URISyntaxException e) {
            // now we give up
        }

        return null;
    }

    @Override
    public InputStream load(
        URI uri,
        boolean noCache
    ) {
        if (uri == null) {
            return null;
        }

        try {
            URL url = uri.toURL();

            if (noCache) {
                URLConnection conn = url.openConnection();
                conn.setUseCaches(false);
                conn.connect(); // explicit for clarity
                return conn.getInputStream();
            } else {
                return url.openStream();
            }
        } catch (IOException ioe) {
            throw new IllegalStateException(ioe.getMessage() + "[uri=" + uri + "]", ioe);
        }
    }

    @Override
    public <S> Iterable<S> services(Class<S> serviceInterface) {
        return ServiceLoader.load(serviceInterface, serviceInterface.getClassLoader());
    }
}
```

b) Run your web app using this system property: -Dnet.time4j.base.ResourceLoader=net.time4j.base.DebugResourceLoader

c) This setup will not solve your issue but very probably gives valuable information what has gone wrong.

**Comment by MenoData** [comment]
I have just now released tzdata-v1.8-2016a. If you use that via gradle 

```gradle
compile ("net.time4j:time4j-tzdata:1.8-2016a")
```
then it should very probably work because it uses as fallback (in case of failed uri-search) the own classloader which should know its own resource files. If you nevertheless experience an `IllegalStateException` then the exception should tell you in detail what has gone wrong in the context of your setup or environment.

Side note: I still consider this fix as not complete but am going to commit further changes to stabilize general resource loading (relevant for i18n- or calendar resources etc.). As soon as v4.13 is out (probably next week) I recommend to use that version.

About the reason why your setup does not work until now, I can only speculate and welcome any informations based on the proposed `DebugResourceLoader`. Background is that Time4J first tries to construct an appropriate URL to find the tz-repository (for speeding up start-time). In standard JDK, Time4J tries to evaluate the protection domain of tz-provider class to construct the URL. My speculation: There is something weird around this protection domain, maybe caused by Tomcat-configuration. Who knows. But the classloader approach should definitely help, otherwise something is really broken in your setup.

Informational: This fix is not relevant for Android (Time4A).

**Comment by MenoData** [comment]
After having stabilized general resource loading, more cannot reasonably be done without further details. The observed NPE will definitely not happen any longer.

**Comment by MenoData** [comment]
I have now released v4.13. Together with tzdata-1.9-2016a, I have set up a standalone Tomcat server (v8.0.32) in a JRE-8.40-environment. No NullPointerException or any other exception was observed when querying the timezones using tzdata-module. Therefore I close this issue.
